Integrate boot file make into top level

mmlb opened the pull request:
## Description

Lift the rules from installer/alpine/Makefile into rules.mk.j2 so that the top-level make can handle building the alpine files itself. Also cleanup/fix the Dockerfile & build.sh files that are responsible for building the Alpine boot files.

## Why is this needed

Extracted out of #72 for better reviewability.


## How Has This Been Tested?
Builds kernel, initramfs, and modloop files as needed.

mmlb committed:
installer/alpine: Indent apk packages in Dockerfile by one level

Just so the actual commands run stand out better.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
installer/alpine: Remove commented out code from Dockerfile

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
installer/alpine: Streamline Dockerfile

Lots of stream line type cleanups here. Things like combining multiple
RUNs into just one, which were split to make use of layer caching while
developing/debugging. Also minimizing layers for changes like `USER` or
`WORKDIR` when `sudo` and `cd` can be done.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
installer/alpine: Do not hardcode aports commit or kernel package info.

This will make for less changes when updating kernel version or flavor.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
installer/alpine: Fetch aports tarball instead of git repo

No need for all of aports history when we only care for a specific commit.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
installer/alpine: Set MAKEFLAGS so abuild will build in parallel

Make build go zoom zoom.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
installer/alpine: Only build one kernel

This will remove any x86_64 configs that are not for the lts kernel from
APKBUILD. The APKBUILD will only build kernels for $ARCH that it has
configs for (makes sense), so by removing non config-lts.x86_64 we won't
waste time building kernels we don't care for.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
rules: Set file type to make with vim modeline magic

For better syntax highlighting mostly.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
make: Move special targets to Makefile

Seems to make more sense in main Makefile instead of rules.mk.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
make: Make use of .DELETE_ON_ERROR GNU extension

This avoids the need to output to .tmp files and later move to $@ after
post processing. Also makes sure that any rules that does
post-processing and doesn't do the .tmp dance is no longer broken
:toot:!

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
installer/alpine: Install eclypsium package in Dockerfile

Not sure why it was being installed in build.sh in the first place. This
now matches how other packages are handled between the Dockerfile and build.sh.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
installer/alpine: Split build.sh into muliptle functions

This will be necessary in a later commit. This way Make will have 3
targets, one for each instead of 1 recipe that builds the artifacts.
Having it in 3 recipes allows the Makefile and make to be trivially
correct, otherwise we have to bend over backwards to handle rules that
generate multiple targets and will likely still have some edge cases
left over. See https://www.gnu.org/software/automake/manual/html_node/Multiple-Outputs.html

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
rules: Build alpine boot files from top level Make

This fixes the Makefile so make knows how to build the alpine boot
files. This closes the loop between boot files and osie packages and now
all of osie is able to be built with one simple `make` that takes into
account all source files.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
rules: Add binfmt_misc deps to targets that need it

This fails much earlier than before which is nice.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
osie-runner: Add saved image file to .dockerignore

Oops, this has been needed for quite a while.

Signed-off-by: Manuel Mendez <manny@packet.com>

DavidHwu commented on installer/alpine/Dockerfile:
```diff
@@ -1,100 +1,101 @@
+    # Setup self-signed keys to sign the built packages \
+    abuild-keygen -a -i -n && \
+    # Fetch alpine package tree \
+    curl -fL https://github.com/alpinelinux/aports/archive/$COMMIT.tar.gz | tar -zxf - && \
```
do we care about failure scenarios ?

mmlb replied:
More than handling via pipefail? https://github.com/tinkerbell/osie/pull/79/files#diff-b67911656ef5d18c4ae36cb6741b7965R2

DavidHwu replied:
My question was geared around what error handling strategy we use... as I see this in docker file I understand now.  Withdrawing.

mmlb committed:
rules: Do not check binfmt_misc in drone

We don't mount the binfmt_misc psudeo-fs in drone so the check always
fails. We could either mount it when needed or just mess with the
rules so it skips it when built in drone. I chose the latter because it
seems to make more sense to have that handled where the deps are defined
in the first place.

Signed-off-by: Manuel Mendez <manny@packet.com>

splaspood commented on rules.mk.j2:
```diff
@@ -220,23 +220,27 @@ else
+endif
 build/osie-aarch64.tar.gz: SED=/FROM/ s|.*|FROM multiarch/ubuntu-debootstrap:arm64-xenial|
-build/osie-x86_64.tar.gz: SED=
+build/osie-x86_64.tar.gz:  SED=
```
extra space, no?

mmlb replied:
lol, @mikemrm said the same thing. I did it on purpose to line up the SEDs in the Makefile, it was annoying me to be unaligned.

splaspood commented on rules.mk.j2:
```diff
@@ -220,23 +220,27 @@ else
+endif
 build/osie-runner-aarch64.tar.gz: SED=/FROM/ s|.*|FROM multiarch/alpine:arm64-v3.7|
-build/osie-runner-x86_64.tar.gz: SED=
+build/osie-runner-x86_64.tar.gz:  SED=
```
another extra space

mmlb replied:
ditto

mmlb committed:
Add freshly built Alpine boot files

Mmmm fresh out the oven and smells delicious.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
Add a way to avoid building Alpine boot files

Since `git` updates the file modified time on clone `make` will want to
rebuild the assets-*/* files. We want to have a way around that in both
ci and for people who may not want to build the files.

So we provide a nice and easy way to skip the building, along with info
in the README and even a very verbose and hopefully eye-catching message
from the Makefile.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
Update README

Updated README contains better, more useful information. Mostly about building OSIE.
Updates to shell.nix and rules.mk.j2 are minor tweaks to keep list of build deps accurate and slim.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb commented:
@splaspood PTAL at the last 2 commits. These setup a way to skip building the Alpine files and adds some updated docs to the README.

mmlb commented:
Trying to figure out why drone/make is still trying to rebuild the alpine assets.

mmlb committed:
Remove qemu-aarch64-static dep from x86_64 target

qemu-aarch64-static is messing with the assets-x86_64 build skipping.
Going to punt on that until a better build system shows up or we
actually build aarch64 assets.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb commented:
Finally got the "skip building alpine assets" stuff correct.

@splaspood @DavidHwu one more looksie please.

mmlb committed:
apps/workflow: Set shebang and shellcheck shell to match other apps

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
apps: Do not loop forever waiting on docker

If docker isn't ready after 15s then something is wrong and might as
well break.

Signed-off-by: Manuel Mendez <manny@packet.com>

mmlb committed:
rules: Ensure separating space after $(Q)/$(E)

Keep seeing weird errors in drone like https://drone.packet.net/tinkerbell/osie/706/9
intermittently. Instead of wasting time fixing/figuring it out just fix
it by avoiding the situation.

Signed-off-by: Manuel Mendez <manny@packet.com>

splaspood commented on apps/runner.sh:
```diff
@@ -86,9 +86,14 @@ phone_home_url=$(jq -r .phone_home_url "$metadata")
-until docker info; do
+i=0
+# shellcheck disable=SC2169
+until docker info >/dev/null || [[ $i -gt 5 ]]; do
```
I feel like this 'wait for docker to be ready' code that's duplicated should be pulled out into a shared function...  but not strongly enough to block the review on.

mmlb replied:
Would agree but this other file would also have to be fetched from /init (really we should just embed them into the initramfs) which I kind of don't really want to do.

splaspood commented on apps/workflow-helper.sh:
```diff
@@ -42,11 +44,14 @@ mkdir -p /etc/docker/ /etc/docker/certs.d/ "/etc/docker/certs.d/$docker_registry
-	echo 'Waiting for docker to respond...'
+i=0
+# shellcheck disable=SC2169
+until docker info >/dev/null || [[ $i -gt 5 ]]; do
```
see above

mmlb replied:
ditto

splaspood commented on apps/osie-installer.sh:
```diff
@@ -135,9 +133,15 @@ cat <<-EOF | grep -v '^\s*$'
+service docker start
+i=0
+# shellcheck disable=SC2169
+until docker info >/dev/null || [[ $i -gt 5 ]]; do
```
see below

splaspood replied:
doh, I figured it put these comments in order of the code, not in the order that I left them...  anyway.. you get the idea ;)

mmlb replied:
heh

splaspood commented on installer/alpine/Dockerfile:
```diff
@@ -1,100 +1,98 @@
+# make sure PKGREL is +1 what is in APKGBUILD
+ENV PKGREL 1
+
+RUN true && \
```
true && ?

mmlb replied:
I want to keep the comment above the command but docker won't let me start a `RUN` with a comment so I just used `true` for that.

splaspood commented on shell.nix:
```diff
@@ -30,14 +30,15 @@ let
     gnused
     libarchive
-    minio
+    minio-client
```
Is this even used anymore?

mmlb replied:
actually yes, we use it to upload to s3. We could use a different tool, but there isn't really a need to change it in this PR.

splaspood replied:
ahh, ok.
